Feature/json factory

lucblender opened the pull request.

lucblender committed:
first draft of factory

lucblender committed:
fixed population of attribute

lucblender committed:
reformat code and add specific exception for the factory

cm0x4D commented on build.gradle:
```diff
@@ -19,6 +19,7 @@ repositories {
 dependencies {
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.10.1'
```
We could update Jackson to 2.14.1

Maybe we take the opportunity to update other libs too, like Paho for example - it is really old.

lucblender replied:
Sure, will take time to test and update all the dependencies

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:
```diff
@@ -226,6 +227,103 @@ public CloudioEndpoint(String uuidOrAppName) throws InvalidUuidException, Invali
     }
 
+    //TODO WIP Here
+    public void addNodesFromJson(String jsonPath) throws CloudioFactoryException {
```
How about to implement the core behaviour of this method with an InputStream as parameter and add a wrapper for the JSON file path. This way a developer could bootstrap the endpoint/the node from any other input stream sources like for example an URL or from inside the jar file.

lucblender replied:
@cm0x4D my idea was to have to have a string as parameter that will be the input the ResourceLoader to get the factory configuration a bit the same as how the certificates are handled. 

Also I thought about adding a parameter to let the user give a factory input without having to code it but still let this method public for more advanced user what do you think?

Since I added the interface to deserialize the factory structure, it's no more relevent to call it addNodesFromJson(String jsonPath). What about addNodesFromFactory(String factoryURI)?

cm0x4D replied:
I am not sure to understand. My idea would be something like this:

```java
public void addNodesFromStream(InputStream nodesInputStream, CloudioFactoryFormat factoryFormat) throws CloudioFactoryException {
        try {
            CloudioFactoryNodes cloudioFactoryNodes = factoryFormat.deserializeNodes(nodesInputStream);
            for (String nodeKey : cloudioFactoryNodes.nodes.keySet()) {
                CloudioFactoryNode cloudioFactoryNode = cloudioFactoryNodes.nodes.get(nodeKey);

                CloudioDynamicNode cloudioDynamicNode;

                if (cloudioFactoryNode.type.equals("CloudioNode")) {
                    cloudioDynamicNode = new CloudioDynamicNode();
                    this.parseCloudioFactoryNode(cloudioFactoryNode, cloudioDynamicNode);
                    System.out.println(nodeKey);

                    this.addNode(nodeKey, cloudioDynamicNode);
                } else {
                    Class<?> nodeClass = Class.forName(cloudioFactoryNode.type);
                    Constructor<?> constructor = nodeClass.getConstructor();
                    Object nodeClassInstance = constructor.newInstance();

                    this.parseCloudioFactoryNode(cloudioFactoryNode, (CloudioDynamicNode) nodeClassInstance);

                    this.addNode(nodeKey, (CloudioNode) nodeClassInstance);
                }
            }
        } catch (Exception exception) {
            throw new CloudioFactoryException(exception);
        }
    }

    public void addNodesFromStream(InputStream nodesInputStream) throws CloudioFactoryException {
        addNodesFromStream(nodesInputStream, this.internal.factoryFormat);
    }

    public void addNodesFromResource(String path) throws CloudioFactoryException {
        try {
            InputStream jsonNodesInputStream = ResourceLoader.getResourceFromLocations(jsonPath,
                    this,
                    "home:" + "/.config/cloud.io/",
                    "file:/etc/cloud.io/",
                    "classpath:cloud.io/");
            addNodesFromStream(jsonNodesInputStream);
        }
        catch (FileNotFoundException exception) {
            throw new CloudioFactoryException("Json file "+jsonPath+" given to add Nodes From Json was not found " +
                    "[\"home:/.config/cloud.io/" + jsonPath + "\", " +
                    "\"file:/etc/cloud.io/" + jsonPath + "\", " +
                    "\"classpath:" + jsonPath + "\"].");
        } catch (Exception exception) {
            throw new CloudioFactoryException(exception);
        }
    }
```

The first method is fully custom, a developer can provide the input stream from where to read the definition and the factory that deserializes it. The second one takes just the input stream and uses the default deserializer by calling the 1st version. The third version creates an input stream from the resource locations and calls the 2nd method.

lucblender replied:
Ooooh alright that's really nice like that, give more possibilities!

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioDynamicNode.java:
```diff
@@ -158,6 +160,23 @@ public CloudioDynamicNode build() {
     private class InternalDynamicNode extends CloudioNode.InternalNode {
         private final NamedItemSet<CloudioObject.InternalObject> dynamicObjects = new NamedItemSet<CloudioObject.InternalObject>();
 
+        public InternalDynamicNode()
```
Oh I see, so we can mix static, annotation objects and dynamic objects - smart!

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/GenericJacksonMessageFormat.java:
```diff
@@ -329,6 +334,14 @@ public byte[] serializeJobsLineOutput(JobsLineOutput jobsLineOutput) {
     }
 
+    @Override
+    public CloudioFactoryNodes deserializeNodes(byte[] data) throws Exception
```
I would not add this here, it has nothing to do with the format of the messages. How about a CloudioFactoryFormat interface and an implementation for JSON.

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioMessageFormat.java:
```diff
@@ -116,4 +116,6 @@ void deserializeLogParameter(byte[] data, LogParameter logParameter)
      */
     byte[] serializeJobsLineOutput(JobsLineOutput jobsLineOutput);
+
+    CloudioFactoryNodes deserializeNodes(byte[] data) throws Exception;
```
See comment in GenericJacksonMessageFormat

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioNode.java:
```diff
@@ -19,7 +19,7 @@ class InternalNode implements CloudioObjectContainer {
         private String name = null;
         private final Set<String> interfaces = new TreeSet<String>();
-        private final NamedItemSet<CloudioObject.InternalObject> objects = new NamedItemSet<CloudioObject.InternalObject>();
+        protected final NamedItemSet<CloudioObject.InternalObject> objects = new NamedItemSet<CloudioObject.InternalObject>();
```
Breaks encapsulation, how about a protected getter method?

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioFactoryObject.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Using
```
@JsonAnySetter
private Map<String, Object> properties = new LinkedHashMap<>();
```
you can automate deserialisation of the additional properties to a map.

lucblender replied:
@cm0x4D I will add those properties map 
Also I was thinking to remove those Json annotation to not make those classes depended of Jackson and configure the object Mapper to ignore the unused fields. What do you think?

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioFactoryNodes.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Using
```
@JsonAnySetter
private Map<String, Object> properties = new LinkedHashMap<>();
```
you can automate deserialisation of the additional properties to a map.

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioFactoryNode.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Using
```
@JsonAnySetter
private Map<String, Object> properties = new LinkedHashMap<>();
```
you can automate deserialisation of the additional properties to a map.

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioFactoryAttribute.java:
```diff
@@ -0,0 +1,10 @@
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Using
```
@JsonAnySetter
private Map<String, Object> properties = new LinkedHashMap<>();
```
you can automate deserialisation of the additional properties to a map.

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:
```diff
@@ -226,6 +227,103 @@ public CloudioEndpoint(String uuidOrAppName) throws InvalidUuidException, Invali
+                    "home:" + "/.config/cloud.io/",
+                    "file:/etc/cloud.io/",
+                    "classpath:cloud.io/");
+            byte[] bytes = jsonNodesInputStream.readAllBytes();
```
AFAIK ObjectMapper can work on input streams too, this would be much more efficient than reading the whole file into an array.

lucblender committed:
dependencies, protected getter, inputstream for deserilize

update dependencies version, add protected getter for objects, use inputstream to deserialize nodes

lucblender committed:
moved deserialization of nodes to specific interface

lucblender committed:
update gitignore, slightly better error handling on factory

lucblender committed:
more possibilities to use factory

MartinMeyer1 committed:
Fixed factory format creation

lucblender committed:
removed any jackson dependency in factory classes

lucblender committed:
removed duplicate code and added factory configurable interface

lucblender committed:
add type for factoryobject and give parameters to node and object

lucblender committed:
some documentation

lucblender committed:
add doc for new methods in CloudioEndpoint

cm0x4D commented on src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java:
```diff
@@ -452,6 +665,8 @@ class InternalEndpoint implements CloudioNodeContainer, MqttCallback, Runnable,
         private static final String MESSAGE_FORMAT                  = "ch.hevs.cloudio.endpoint.messageFormat";
         private static final String MESSAGE_FORMAT_DEFAULT          = "CBOR";
+        private static final String FACTORY_FORMAT                  = "ch.hevs.cloudio.endpoint.factoryFormat";
+        private static final String MESSAGE_FACTORY_DEFAULT          = "JSON";
```
FACTORY_FORMAT_DEFAULT ?

lucblender replied:
Fixed!

cm0x4D replied:
👍👍👍

lucblender committed:
fixe typo related to factory

MartinMeyer1 committed:
Added source and doc publication
